clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -2882,6 +2882,82 @@ raw2trace_t::set_instr_summary_flags(raw2trace_thread_data_t *tdata, uint64 modi
     return true;
 }
 
+#if defined(AARCH64)
+/* TODO i#5365: append_bb_entries() takes the size of the scatter/gather memory operand
+ * to be the per-element value and uses that to create the read/write memref entries.
+ * The AArch64 IR currently uses the maximum amount of data transferred by the instruction
+ * (number of elements * per element size) instead so until we change the codec to use the
```

Comment:
Hmm I think x86 also has the max size -- right @abhinav92003?  I see `OP_vpscatterdd, 0x6638a008, catSIMD, "vpscatterdd", MVd, KEw, KEw, Ve, xx, mrm|evex|reqp|ttnone|nok0, x, END_LIST},` and `#define Mvd TYPE_M, OPSZ_16_vex32`.  So what happens today for x86 for the per record size?

Reply:
Looking at the trace for clients/drcachesim/tests/allasm_scattergather_x86.asm, each mem accesses has the per-element size:

```
          56          45:     1393700 ifetch      10 byte(s) @ 0x0000000000401124 c4 22 11 90 24 9d 39 vpgatherdd 0x00402039(,%xmm11,4), %xmm13, %xmm12
          56          45:     1393700                                             20 40 00
          57          45:     1393700 read         4 byte(s) @ 0x000000000040203d by PC 0x0000000000401124
          58          45:     1393700 read         4 byte(s) @ 0x0000000000402045 by PC 0x0000000000401124
          59          45:     1393700 read         4 byte(s) @ 0x0000000000402041 by PC 0x0000000000401124
```

As linked in my other comment, I saw a different definition for `MVd`, which made sense so I didn't look further.

```
#define MVd TYPE_VSIB, OPSZ_4
```

but looks like there are two definitions. Not sure how that works.

Reply:
I am looking into why we don't get a warning on the re-definition, and whether this is making some non-scatter instructions incorrectly have OPSZ_4 as their size (we only want the element size for scatter/gather or rep-string -- regular SIMD should have the total size).

Reply:
Ah, mystery solved: it's `MVd` vs `Mvd`: the case of the `v` is different.  (My emacs does a case-insensitive interactive search.)

Reply:
> we only want the element size for scatter/gather or rep-string -- regular SIMD should have the total size

What about predicated contiguous load/store instructions? I'm not sure there is a direct equivalent in x86. Because they have a predicate mask like the scatter/gather instructions we treat them like scatter/gather instructions in a lot of places in DynamoRIO so I would suggest we keep them consistent with scatter/gather instructions and use the per-element size.

Reply:
There are mask registers in AVX-512 which I think apply to many operations and not just scatter/gather so I think the same concept does exist there.

Hmm, it's seeming like we want dr_opnd_query_flags_t for the size now.  For predicated contiguous, a taint-tracking tool (such as Dr. Memory) might want the max size for loads when checking taint bits (on a fastpath anyway) but would have to loop over the per-element for stores when setting taint bits.  Hmm.  I guess the slowpath would loop too.  Maybe file an issue on this predicated size problem covering all platforms, and go ahead w/ your current plan for now?

This is similar to the half-register and holes-in-register complex SIMD interleaving operations: xref #1382, #6218.

Xref #5638.

Reply:
Thanks. Created #6561